### PR TITLE
Closes #1480: Made fail fast solution for calculating file size limit.

### LINKED ISSUE #1480: When transfering a file via API, its size is evaluated after the transfer is completed
So it is in fact possible to send a large file and clog the temporary storage

### PR DESCRIPTION
**Pull request by MrK191**

### CODE REVIEW

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DataFileServiceBean.java [hunk 1]
```diff
@@ -1590,23 +1568,31 @@ private File saveInputStreamInTempFile(InputStream inputStream, Long fileSizeLim
...
 
-            return tempFile.toFile();
+                throw new FileExceedsMaxSizeException(MessageFormat.format(BundleUtil.getStringFromBundle("file.addreplace.error.file_exceeds_limit"),
+                                                                           bytesToHumanReadable(fileSizeLimit)));
+            }
+            outStream.write(buffer, 0, bytesRead);
         }
-        throw new IOException("Failed to save uploaded file.");
+        inputStream.close();
+        outStream.close();
```
Comment (wtr-icm) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DataFileServiceBean.java]: Could we modify the flow here in order to have `outStream` declared inside `try-with-resources` clause? If not then we should at least place all `close()`s inside `try-finally` blocks.

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DataFileServiceBean.java [hunk 2]
```diff
@@ -1590,23 +1568,31 @@ private File saveInputStreamInTempFile(InputStream inputStream, Long fileSizeLim
...
+        byte[] buffer = new byte[8 * 1024];
+        int bytesRead;
+        long totalBytesRead = 0;
+        while ((bytesRead = inputStream.read(buffer)) != -1) {
+            totalBytesRead += bytesRead;
+            if (fileSizeLimit != null && totalBytesRead > fileSizeLimit) {
                 try {
                     tempFile.toFile().delete();
                 } catch (Exception ex) {
                 }
```
Comment (wtr-icm) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DataFileServiceBean.java]: We should log that exception at least.

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DataFileServiceBean.java [hunk 3]
```diff
@@ -1590,23 +1568,31 @@ private File saveInputStreamInTempFile(InputStream inputStream, Long fileSizeLim
...
-            Long fileSize = tempFile.toFile().length();
-            if (fileSizeLimit != null && fileSize > fileSizeLimit) {
+        byte[] buffer = new byte[8 * 1024];
+        int bytesRead;
+        long totalBytesRead = 0;
+        while ((bytesRead = inputStream.read(buffer)) != -1) {
+            totalBytesRead += bytesRead;
+            if (fileSizeLimit != null && totalBytesRead > fileSizeLimit) {
                 try {
                     tempFile.toFile().delete();
```
Comment (wtr-icm) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DataFileServiceBean.java]: I'm not sure, but it seems to me that we should try to close out stream before deletion.

**Review by wtr-icm — CHANGES_REQUESTED** [review]

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DataFileServiceBean.java (hunk 3 above)
Comment (MrK191) [reply dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DataFileServiceBean.java]: I think closing it in finally clause is sufficient.

Comment (wtr-icm) [reply dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DataFileServiceBean.java]: I thought not about sufficiency but about something different: if file is deleted while we have open stream to that file, then can we safely close that stream (ie. without exception)?
But I think it should be ok.

**Review by wtr-icm — APPROVED** [review]
I see that your automatic code formatter behaves in a bit strange way, especially in this file.

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DataFileServiceBean.java (hunk 3 above)
Comment (MrK191) [reply dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DataFileServiceBean.java]: Finally clause is executed even if exception is thrown inside try block

**Pull request merged**